pkg/metrics/metrics.go

```diff
@@ -0,0 +1,158 @@
+//
+// Copyright (c) 2018 The heketi Authors
+//
+// This file is licensed to you under your choice of the GNU Lesser
+// General Public License, version 3 or any later version (LGPLv3 or
+// later), or the GNU General Public License, version 2 (GPLv2), in all
+// cases as published by the Free Software Foundation.
+//
+
+package metrics
+
+import (
+	"github.com/heketi/heketi/apps"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	"net/http"
+)
+
+type Metrics struct {
+	app apps.Application
+}
+
+const (
+	namespace = "heketi"
+)
+
+var (
+	up = promDesc(
+		"up",
+		"Was the last query of heketi successful.",
+		nil,
+	)
+
+	clusterCount = promDesc(
+		"cluster_count",
+		"Number of clusters at last query.",
+		nil,
+	)
+
+	volumesCount = promDesc(
+		"volumes_count",
+		"How many volumes were up at the last query.",
+		[]string{"cluster"},
+	)
+
+	nodesCount = promDesc(
+		"nodes_count",
+		"How many Nodes were up at the last query.",
+		[]string{"cluster"},
+	)
+
+	deviceCount = promDesc(
+		"device_count",
+		"How many Devices were up at the last query.",
+		[]string{"cluster", "hostname"},
+	)
+
+	deviceSize = promDesc(
+		"device_size",
+		"How many Devices were up at the last query.",
+		[]string{"cluster", "hostname", "device"},
+	)
+
+	deviceFree = promDesc(
+		"device_free",
+		"How many Devices were up at the last query.",
+		[]string{"cluster", "hostname", "device"},
+	)
+
+	deviceUsed = promDesc(
+		"device_used",
+		"How many Devices were up at the last query.",
+		[]string{"cluster", "hostname", "device"},
+	)
+
+	brickCount = promDesc(
+		"brick_count",
+		"Number of bricks at last query.",
+		[]string{"cluster", "hostname", "device"},
+	)
+)
+
+func promDesc(name, help string, variableLabels []string) *prometheus.Desc {
+	return prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, "", name),
+		help,
+		variableLabels,
+		nil,
+	)
+}
+
+// Describe all the metrics exported by Heketi exporter. It implements prometheus.Collector.
+func (m *Metrics) Describe(ch chan<- *prometheus.Desc) {
+	ch <- clusterCount
+	ch <- volumesCount
+	ch <- nodesCount
+	ch <- deviceCount
+	ch <- deviceSize
+	ch <- deviceFree
+	ch <- deviceUsed
+	ch <- brickCount
+
+}
+
+// Collect metrics from heketi app
+func (m *Metrics) Collect(ch chan<- prometheus.Metric) {
+	topinfo, err := m.app.TopologyInfo()
+	if err != nil {
+		ch <- prometheus.MustNewConstMetric(
+			up, prometheus.GaugeValue, 0.0,
+		)
+	} else {
+		ch <- prometheus.MustNewConstMetric(
+			up, prometheus.GaugeValue, 1.0,
+		)
+	}
```

Comment:
Since most of this block is shared this could read better as:
```golang
upVal := 0.0
if err == nil {
    upVal := 1.0
}
ch <- prometheus.MustNewConstMetric(up, prometheus.GaugeValue, upVal)

```